cmd/standby-forecaster/main.go

```diff
@@ -0,0 +1,399 @@
+		totalServersNeeded = int(math.Ceil(forecastValue / config.MetricToServerConversionOperationValue))
+	} else if config.MetricToServerConversionOperation == "multiply" {
+		totalServersNeeded = int(math.Ceil(forecastValue * config.MetricToServerConversionOperationValue))
+	} else {
+		log.Warn("Invalid metric to server conversion operation setting. No conversion will occur", "operation", config.MetricToServerConversionOperation)
+		totalServersNeeded = int(forecastValue)
+	}
+	log.Debug("Total servers needed", "servers", totalServersNeeded)
+	TotalServersForecast.Set(float64(totalServersNeeded))
+	return totalServersNeeded, nil
+}
+
+// get kube client using the service account token for the pod
+func getKubeClient(namespace string, runInCluster bool, kubeConfigPath string) (*GameServerBuildClient, error) {
+
+	var kubeConfig *rest.Config
+	var err error
+	if runInCluster {
+		kubeConfig, err = rest.InClusterConfig()
+	} else {
+		var configBytes []byte
+		configBytes, err = ioutil.ReadFile(kubeConfigPath)
+		if err != nil {
+			return nil, err
+		}
+		kubeConfig, err = clientcmd.RESTConfigFromKubeConfig(configBytes)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	config := *kubeConfig
+	config.ContentConfig.GroupVersion = &schema.GroupVersion{Group: "mps.playfab.com", Version: "v1alpha1"}
+	config.APIPath = "/apis"
+	config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+	config.UserAgent = rest.DefaultKubernetesUserAgent()
+	client, err := rest.RESTClientFor(&config)
+	if err != nil {
+		return nil, err
+	}
+	return &GameServerBuildClient{restClient: client, ns: namespace}, nil
+}
+
+func getLinearRegressionPrediction(data []timeSeriesPoint, timeToForcast int64) (float64, error) {
+	// create the regression model
+	r := new(regression.Regression)
+
+	// loop through the data
+	for i := 0; i < len(data); i++ {
+		r.Train(regression.DataPoint(data[i].value, []float64{float64(data[i].time.Unix())}))
+	}
+	err := r.Run()
+	if err != nil {
+		return 0.0, err
+	}
+	// predict the value
+	return r.Predict([]float64{float64(timeToForcast)})
+}
+
+func queryPrometheusMetrics(ctx context.Context, queryUrl, metricQuery string, historyRange time.Duration) ([]timeSeriesPoint, error) {
+	// create a prometheus client
+	client, err := promClient.NewClient(promClient.Config{
+		Address: queryUrl,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	api := promapi.NewAPI(client)
+
+	// query the prometheus metrics
+	rng := promapi.Range{}
+	rng.Start = time.Now().Add(-historyRange)
+	rng.End = time.Now()
+	rng.Step = time.Minute
+	queryCtx, queryCancel := context.WithTimeout(ctx, time.Duration(30*time.Second))
+	defer queryCancel()
+	queryResult, _, err := api.QueryRange(queryCtx, metricQuery, rng)
+	if err != nil {
+		return nil, err
+	}
+	results := queryResult.(model.Matrix)
+	// initialize the time series
+	timeSeries := make([]timeSeriesPoint, 0)
+
+	// loop through the values
+	for _, point := range results[0].Values {
```

Comment:
got a panic after some time in this line, single GameServerBuild with 2 standingBy and 1 active:
```
standby-forecaster-5bd4868487-sgr8t    0/1     CrashLoopBackOff   3 (25s ago)   2m16s
dgkanatsios@desktopdigkanat:thundernetes$ k logs standby-forecaster-5bd4868487-sgr8t 
time="2022-07-18T22:11:57Z" level=info msg="Starting Standby Forecaster"
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
main.queryPrometheusMetrics({0x187fb88, 0xc00009ab00}, {0xc000123bf0?, 0x0?}, {0xc00014ed68, 0x18}, 0x13a52453c000)
        /workspace/cmd/standby-forecaster/main.go:371 +0x590
main.main()
        /workspace/cmd/standby-forecaster/main.go:137 +0x527
```